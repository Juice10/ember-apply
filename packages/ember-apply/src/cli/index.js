#!/usr/bin/env node
// @ts-check
import assert from 'assert';
import path, { dirname } from 'path';
import { fileURLToPath } from 'url';
import yargs from 'yargs';
import chalk from 'chalk';
import { hideBin } from 'yargs/helpers';

/**
 * @typedef {object} Options
 * @property {string} [ name ]
 * @property {boolean} [ verbose ]
 *
 */

const __dirname = dirname(fileURLToPath(import.meta.url));

process.on('uncaughtException', function (err) {
  console.error(chalk.red(err));
});

yargs(hideBin(process.argv))
  .command(
    '$0 <name>',
    'Apply a feature to the current project',
    (yargs) => {
      return yargs.positional('name', {
        describe: 'Name of the feature to apply',
        type: 'string',
      });
    },
    async (argv) => {
      /** @type {Options} */
      let args = argv;

      assert(args.name, 'name is required');

      const applyable = await getApplyable(args);

      assert(applyable, 'Could not find an applyable feature. Does it have a default export?');
      assert(typeof applyable === 'function', 'applyable must be a function');

      await applyable();
    }
  )
  .option('verbose', {
    alias: 'v',
    type: 'boolean',
    description: 'Run with verbose logging',
  })
  .demandCommand(1)
  .parse();

/**
 * @param {Options} options
 *
 */
async function getApplyable(options) {
  let applyableModule = await resolveApplyable(options);

  assert(applyableModule, 'Could not find an applyable feature. Does it have a default export?');

  return applyableModule.default;
}

/**
 * @param {Options} options
 */
async function resolveApplyable(options) {
  return (
    // Could be a local path on the file system
    (await resolvePath(options)) ||
    // Could be a npm package
    (await resolvePackage(options))
  );
}

/**
 * @param {Options} options
 */
async function resolvePackage(options) {
  let { name } = options;

  // TODO: prompt user before running this code
  //       (any package can be placed here)

  // https://www.skypack.dev/view/@ember-apply/tailwind
  // import emberApplyTailwind from 'https://cdn.skypack.dev/@ember-apply/tailwind';
  return await tryResolve(`https://cdn.skypack.dev/${name}`, options);
}

/**
 * @param {Options} options
 */
async function resolvePath(options) {
  let cwd = process.cwd();
  let { name, verbose } = options;

  assert(name, 'name is required');

  return (
    // local, but specified index.js
    (await tryResolve(path.join(cwd, name), options)) ||
    // local, but without index.js
    (await tryResolve(path.join(cwd, name, 'index.js'), options)) ||
    // local, but absolute path
    (await tryResolve(path.join(name), options)) ||
    // local, but absolute path without specifying index.js
    (await tryResolve(path.join(name, 'index.js'), options))
  );
}

/**
 * @param {string} url - the path to import
 * @param {Options} [options]
 */
async function tryResolve(url, options = {}) {
  try {
    if (options.verbose) {
      console.info(chalk.gray(`Checking ${url}`));
    }

    return await import(url);
  } catch (error) {
    if (options.verbose) {
      console.error(chalk.red(error));
    }

    return;
  }
}
