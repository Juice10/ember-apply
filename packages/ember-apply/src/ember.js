// // @ts-check

// import { transformTemplate } from './-private/transform-templates.js';

// /**
//  * ember-specific utilities
//  *
//  * See [emberjs.com](https://emberjs.com/) for more information on ember.
//  *
//  */
// export const ember = {
//   transformTemplate,
// };
