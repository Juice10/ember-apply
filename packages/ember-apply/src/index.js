// @ts-check

export { addScript, addScripts, addDevDependencies, modify } from './-private/package-json.js';
export { gitIgnore } from './-private/project-utils.js';

export { transformHTML, addHTML } from './-private/transform-html.js';

export { applyFolder, copyFileTo } from './-private/file-utils.js';

export { transformScript } from './-private/transform-js.js';

export { transformTemplate } from './-private/transform-templates.js';
