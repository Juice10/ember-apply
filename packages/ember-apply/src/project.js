// // @ts-check

// import { gitIgnore, addScript, addScripts, addDevDependencies } from './-private/project-utils.js';
// import { applyFolder, copyFileTo } from './-private/file-utils.js';

// /**
//  * project utilities for manipulating package.json and the overall project
//  */
// export const project = {
//   addScript,
//   addScripts,
//   gitIgnore,
//   applyFolder,
//   copyFileTo,
//   addDevDependencies,
// };
