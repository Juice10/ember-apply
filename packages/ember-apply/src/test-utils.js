import fs from 'fs/promises';
import path, { dirname } from 'path';
import os from 'os';
import { fileURLToPath } from 'url';
import { execa } from 'execa';

const __dirname = dirname(fileURLToPath(import.meta.url));

export async function newEmberApp() {
  let dir = await fs.mkdtemp(path.join(os.tmpdir(), 'ember-apply-test-app--'));

  await execa('ember', ['new', 'test-app', '--skip-npm'], {
    cwd: dir,
  });

  return path.join(dir, 'test-app');
}

/**
 * @param {string} appPath the app to apply the applyable to
 * @param {string} applyablePath the path to the applyable
 */
export async function apply(appPath, applyablePath) {
  let cli = path.join(__dirname, 'cli/index.js');
  let target = path.resolve(applyablePath);

  await execa('node', [cli, target], {
    cwd: appPath,
  });
}

/**
 * @param {string} appPath the path to git diff
 */
export async function diffSummary(appPath) {
  let { stdout } = await execa('git', ['diff', '--compact-summary'], {
    cwd: appPath,
  });

  return stdout;
}

/**
 * @param {string} appPath the path to git diff
 */
export async function diff(appPath) {
  let { stdout } = await execa('git', ['diff'], {
    cwd: appPath,
  });

  return stdout;
}
