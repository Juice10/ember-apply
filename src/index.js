// @ts-check
export {
  transformScript,
  transformTemplate,
  transformHTML,
  addHTML,
  gitIgnore,
  addScript,
} from './transform-utils.js';

export { applyFolder, copyFileTo } from './file-utils.js';
