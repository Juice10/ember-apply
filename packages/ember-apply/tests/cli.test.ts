import path from 'path';
import { describe, expect, it } from 'vitest';

import { apply, diff, diffSummary, newEmberApp } from '../src/test-utils';

describe('CLI', () => {
  describe('default command', () => {
    describe('using Tailwind as an example', () => {
      it.skip('package: @ember-apply/tailwind', async () => {
        let appLocation = await newEmberApp();

        await apply(appLocation, '@ember-apply/tailwind');

        expect(await diffSummary(appLocation)).toMatchSnapshot();
        expect(await diff(appLocation)).toMatchSnapshot();
      }, 60_000);

      it('local relative path', async () => {
        let appLocation = await newEmberApp();

        await apply(appLocation, '../ember/tailwind');

        expect(await diffSummary(appLocation)).toMatchSnapshot();
        expect(await diff(appLocation)).toMatchSnapshot();
      });

      it('local relative path (with index)', async () => {
        let appLocation = await newEmberApp();

        await apply(appLocation, '../ember/tailwind/index.js');

        expect(await diffSummary(appLocation)).toMatchSnapshot();
        expect(await diff(appLocation)).toMatchSnapshot();
      });

      it('local absolute path', async () => {
        let appLocation = await newEmberApp();
        let target = path.resolve('../ember/tailwind');

        await apply(appLocation, target);

        expect(await diffSummary(appLocation)).toMatchSnapshot();
        expect(await diff(appLocation)).toMatchSnapshot();
      });

      it('local absolute path (with index)', async () => {
        let appLocation = await newEmberApp();
        let target = path.resolve('../ember/tailwind/index.js');

        await apply(appLocation, target);

        expect(await diffSummary(appLocation)).toMatchSnapshot();
        expect(await diff(appLocation)).toMatchSnapshot();
      });
    });
  });
});
