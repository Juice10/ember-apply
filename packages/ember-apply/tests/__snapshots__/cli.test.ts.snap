// Vitest Snapshot v1

exports[`CLI > default command > using Tailwind as an example > local absolute path (with index) 1`] = `
" .gitignore       |  3 +++
 app/index.html   | 19 +++++++++----------
 package.json     | 11 ++++++++---
 tests/index.html | 21 ++++++++++-----------
 4 files changed, 30 insertions(+), 24 deletions(-)"
`;

exports[`CLI > default command > using Tailwind as an example > local absolute path (with index) 2`] = `
"diff --git a/.gitignore b/.gitignore
index 9bab45a..c4e2f3e 100644
--- a/.gitignore
+++ b/.gitignore
@@ -1,6 +1,9 @@
 # See https://help.github.com/ignore-files/ for more about ignoring files.
 
+
 # compiled output
+public/assets/tailwind.css
+
 /dist/
 /tmp/
 
diff --git a/app/index.html b/app/index.html
index 2398775..f038aae 100644
--- a/app/index.html
+++ b/app/index.html
@@ -1,25 +1,24 @@
-<!DOCTYPE html>
-<html>
-  <head>
+<!DOCTYPE html><html><head>
     <meta charset=\\"utf-8\\">
     <meta http-equiv=\\"X-UA-Compatible\\" content=\\"IE=edge\\">
-    <title>TestApp</title>
+    <title></title>
     <meta name=\\"description\\" content=\\"\\">
     <meta name=\\"viewport\\" content=\\"width=device-width, initial-scale=1\\">
 
-    {{content-for \\"head\\"}}
+    </head><body>{{content-for \\"head\\"}}
 
-    <link integrity=\\"\\" rel=\\"stylesheet\\" href=\\"{{rootURL}}assets/vendor.css\\">
+    <link integrity=\\"\\" rel=\\"stylesheet\\" href=\\"{{rootURL}}assets/tailwind.css\\"><link integrity=\\"\\" rel=\\"stylesheet\\" href=\\"{{rootURL}}assets/vendor.css\\">
     <link integrity=\\"\\" rel=\\"stylesheet\\" href=\\"{{rootURL}}assets/test-app.css\\">
 
     {{content-for \\"head-footer\\"}}
-  </head>
-  <body>
+  
+  
     {{content-for \\"body\\"}}
 
     <script src=\\"{{rootURL}}assets/vendor.js\\"></script>
     <script src=\\"{{rootURL}}assets/test-app.js\\"></script>
 
     {{content-for \\"body-footer\\"}}
-  </body>
-</html>
+  
+
+</body></html>
\\\\ No newline at end of file
diff --git a/package.json b/package.json
index 976763c..e14a6d4 100644
--- a/package.json
+++ b/package.json
@@ -11,7 +11,7 @@
     \\"test\\": \\"tests\\"
   },
   \\"scripts\\": {
-    \\"build\\": \\"ember build --environment=production\\",
+    \\"build\\": \\"npm run tailwind:build && ember build --environment=production\\",
     \\"lint\\": \\"npm-run-all --aggregate-output --continue-on-error --parallel \\\\\\"lint:!(fix)\\\\\\"\\",
     \\"lint:fix\\": \\"npm-run-all --aggregate-output --continue-on-error --parallel lint:*:fix\\",
     \\"lint:hbs\\": \\"ember-template-lint .\\",
@@ -20,7 +20,9 @@
     \\"lint:js:fix\\": \\"eslint . --fix\\",
     \\"start\\": \\"ember serve\\",
     \\"test\\": \\"npm-run-all lint test:*\\",
-    \\"test:ember\\": \\"ember test\\"
+    \\"test:ember\\": \\"ember test\\",
+    \\"tailwind:build\\": \\"npx tailwindcss -i ./tailwind-input.css -o ./public/assets/tailwind.css\\",
+    \\"tailwind:watch\\": \\"npx tailwindcss -i ./tailwind-input.css -o ./public/assets/tailwind.css --watch\\"
   },
   \\"devDependencies\\": {
     \\"@ember/optional-features\\": \\"^2.0.0\\",
@@ -59,7 +61,10 @@
     \\"prettier\\": \\"^2.5.1\\",
     \\"qunit\\": \\"^2.17.2\\",
     \\"qunit-dom\\": \\"^2.0.0\\",
-    \\"webpack\\": \\"^5.65.0\\"
+    \\"webpack\\": \\"^5.65.0\\",
+    \\"autoprefixer\\": \\"^10.0.0\\",
+    \\"postcss\\": \\"^8.0.0\\",
+    \\"tailwindcss\\": \\"^3.0.0\\"
   },
   \\"engines\\": {
     \\"node\\": \\"12.* || 14.* || >= 16\\"
diff --git a/tests/index.html b/tests/index.html
index ec6650a..b61f82e 100644
--- a/tests/index.html
+++ b/tests/index.html
@@ -1,23 +1,21 @@
-<!DOCTYPE html>
-<html>
-  <head>
+<!DOCTYPE html><html><head>
     <meta charset=\\"utf-8\\">
     <meta http-equiv=\\"X-UA-Compatible\\" content=\\"IE=edge\\">
-    <title>TestApp Tests</title>
+    <title></title>
     <meta name=\\"description\\" content=\\"\\">
     <meta name=\\"viewport\\" content=\\"width=device-width, initial-scale=1\\">
 
-    {{content-for \\"head\\"}}
+    </head><body>{{content-for \\"head\\"}}
     {{content-for \\"test-head\\"}}
 
-    <link rel=\\"stylesheet\\" href=\\"{{rootURL}}assets/vendor.css\\">
+    <link rel=\\"stylesheet\\" href=\\"{{rootURL}}assets/tailwind.css\\"><link rel=\\"stylesheet\\" href=\\"{{rootURL}}assets/vendor.css\\">
     <link rel=\\"stylesheet\\" href=\\"{{rootURL}}assets/test-app.css\\">
     <link rel=\\"stylesheet\\" href=\\"{{rootURL}}assets/test-support.css\\">
 
     {{content-for \\"head-footer\\"}}
     {{content-for \\"test-head-footer\\"}}
-  </head>
-  <body>
+  
+  
     {{content-for \\"body\\"}}
     {{content-for \\"test-body\\"}}
 
@@ -28,7 +26,7 @@
       </div>
     </div>
 
-    <script src=\\"/testem.js\\" integrity=\\"\\" data-embroider-ignore></script>
+    <script src=\\"/testem.js\\" integrity=\\"\\" data-embroider-ignore=\\"\\"></script>
     <script src=\\"{{rootURL}}assets/vendor.js\\"></script>
     <script src=\\"{{rootURL}}assets/test-support.js\\"></script>
     <script src=\\"{{rootURL}}assets/test-app.js\\"></script>
@@ -36,5 +34,6 @@
 
     {{content-for \\"body-footer\\"}}
     {{content-for \\"test-body-footer\\"}}
-  </body>
-</html>
+  
+
+</body></html>
\\\\ No newline at end of file"
`;

exports[`CLI > default command > using Tailwind as an example > local absolute path 1`] = `
" .gitignore       |  3 +++
 app/index.html   | 19 +++++++++----------
 package.json     | 11 ++++++++---
 tests/index.html | 21 ++++++++++-----------
 4 files changed, 30 insertions(+), 24 deletions(-)"
`;

exports[`CLI > default command > using Tailwind as an example > local absolute path 2`] = `
"diff --git a/.gitignore b/.gitignore
index 9bab45a..c4e2f3e 100644
--- a/.gitignore
+++ b/.gitignore
@@ -1,6 +1,9 @@
 # See https://help.github.com/ignore-files/ for more about ignoring files.
 
+
 # compiled output
+public/assets/tailwind.css
+
 /dist/
 /tmp/
 
diff --git a/app/index.html b/app/index.html
index 2398775..f038aae 100644
--- a/app/index.html
+++ b/app/index.html
@@ -1,25 +1,24 @@
-<!DOCTYPE html>
-<html>
-  <head>
+<!DOCTYPE html><html><head>
     <meta charset=\\"utf-8\\">
     <meta http-equiv=\\"X-UA-Compatible\\" content=\\"IE=edge\\">
-    <title>TestApp</title>
+    <title></title>
     <meta name=\\"description\\" content=\\"\\">
     <meta name=\\"viewport\\" content=\\"width=device-width, initial-scale=1\\">
 
-    {{content-for \\"head\\"}}
+    </head><body>{{content-for \\"head\\"}}
 
-    <link integrity=\\"\\" rel=\\"stylesheet\\" href=\\"{{rootURL}}assets/vendor.css\\">
+    <link integrity=\\"\\" rel=\\"stylesheet\\" href=\\"{{rootURL}}assets/tailwind.css\\"><link integrity=\\"\\" rel=\\"stylesheet\\" href=\\"{{rootURL}}assets/vendor.css\\">
     <link integrity=\\"\\" rel=\\"stylesheet\\" href=\\"{{rootURL}}assets/test-app.css\\">
 
     {{content-for \\"head-footer\\"}}
-  </head>
-  <body>
+  
+  
     {{content-for \\"body\\"}}
 
     <script src=\\"{{rootURL}}assets/vendor.js\\"></script>
     <script src=\\"{{rootURL}}assets/test-app.js\\"></script>
 
     {{content-for \\"body-footer\\"}}
-  </body>
-</html>
+  
+
+</body></html>
\\\\ No newline at end of file
diff --git a/package.json b/package.json
index 976763c..e14a6d4 100644
--- a/package.json
+++ b/package.json
@@ -11,7 +11,7 @@
     \\"test\\": \\"tests\\"
   },
   \\"scripts\\": {
-    \\"build\\": \\"ember build --environment=production\\",
+    \\"build\\": \\"npm run tailwind:build && ember build --environment=production\\",
     \\"lint\\": \\"npm-run-all --aggregate-output --continue-on-error --parallel \\\\\\"lint:!(fix)\\\\\\"\\",
     \\"lint:fix\\": \\"npm-run-all --aggregate-output --continue-on-error --parallel lint:*:fix\\",
     \\"lint:hbs\\": \\"ember-template-lint .\\",
@@ -20,7 +20,9 @@
     \\"lint:js:fix\\": \\"eslint . --fix\\",
     \\"start\\": \\"ember serve\\",
     \\"test\\": \\"npm-run-all lint test:*\\",
-    \\"test:ember\\": \\"ember test\\"
+    \\"test:ember\\": \\"ember test\\",
+    \\"tailwind:build\\": \\"npx tailwindcss -i ./tailwind-input.css -o ./public/assets/tailwind.css\\",
+    \\"tailwind:watch\\": \\"npx tailwindcss -i ./tailwind-input.css -o ./public/assets/tailwind.css --watch\\"
   },
   \\"devDependencies\\": {
     \\"@ember/optional-features\\": \\"^2.0.0\\",
@@ -59,7 +61,10 @@
     \\"prettier\\": \\"^2.5.1\\",
     \\"qunit\\": \\"^2.17.2\\",
     \\"qunit-dom\\": \\"^2.0.0\\",
-    \\"webpack\\": \\"^5.65.0\\"
+    \\"webpack\\": \\"^5.65.0\\",
+    \\"autoprefixer\\": \\"^10.0.0\\",
+    \\"postcss\\": \\"^8.0.0\\",
+    \\"tailwindcss\\": \\"^3.0.0\\"
   },
   \\"engines\\": {
     \\"node\\": \\"12.* || 14.* || >= 16\\"
diff --git a/tests/index.html b/tests/index.html
index ec6650a..b61f82e 100644
--- a/tests/index.html
+++ b/tests/index.html
@@ -1,23 +1,21 @@
-<!DOCTYPE html>
-<html>
-  <head>
+<!DOCTYPE html><html><head>
     <meta charset=\\"utf-8\\">
     <meta http-equiv=\\"X-UA-Compatible\\" content=\\"IE=edge\\">
-    <title>TestApp Tests</title>
+    <title></title>
     <meta name=\\"description\\" content=\\"\\">
     <meta name=\\"viewport\\" content=\\"width=device-width, initial-scale=1\\">
 
-    {{content-for \\"head\\"}}
+    </head><body>{{content-for \\"head\\"}}
     {{content-for \\"test-head\\"}}
 
-    <link rel=\\"stylesheet\\" href=\\"{{rootURL}}assets/vendor.css\\">
+    <link rel=\\"stylesheet\\" href=\\"{{rootURL}}assets/tailwind.css\\"><link rel=\\"stylesheet\\" href=\\"{{rootURL}}assets/vendor.css\\">
     <link rel=\\"stylesheet\\" href=\\"{{rootURL}}assets/test-app.css\\">
     <link rel=\\"stylesheet\\" href=\\"{{rootURL}}assets/test-support.css\\">
 
     {{content-for \\"head-footer\\"}}
     {{content-for \\"test-head-footer\\"}}
-  </head>
-  <body>
+  
+  
     {{content-for \\"body\\"}}
     {{content-for \\"test-body\\"}}
 
@@ -28,7 +26,7 @@
       </div>
     </div>
 
-    <script src=\\"/testem.js\\" integrity=\\"\\" data-embroider-ignore></script>
+    <script src=\\"/testem.js\\" integrity=\\"\\" data-embroider-ignore=\\"\\"></script>
     <script src=\\"{{rootURL}}assets/vendor.js\\"></script>
     <script src=\\"{{rootURL}}assets/test-support.js\\"></script>
     <script src=\\"{{rootURL}}assets/test-app.js\\"></script>
@@ -36,5 +34,6 @@
 
     {{content-for \\"body-footer\\"}}
     {{content-for \\"test-body-footer\\"}}
-  </body>
-</html>
+  
+
+</body></html>
\\\\ No newline at end of file"
`;

exports[`CLI > default command > using Tailwind as an example > local relative path (with index) 1`] = `
" .gitignore       |  3 +++
 app/index.html   | 19 +++++++++----------
 package.json     | 11 ++++++++---
 tests/index.html | 21 ++++++++++-----------
 4 files changed, 30 insertions(+), 24 deletions(-)"
`;

exports[`CLI > default command > using Tailwind as an example > local relative path (with index) 2`] = `
"diff --git a/.gitignore b/.gitignore
index 9bab45a..c4e2f3e 100644
--- a/.gitignore
+++ b/.gitignore
@@ -1,6 +1,9 @@
 # See https://help.github.com/ignore-files/ for more about ignoring files.
 
+
 # compiled output
+public/assets/tailwind.css
+
 /dist/
 /tmp/
 
diff --git a/app/index.html b/app/index.html
index 2398775..f038aae 100644
--- a/app/index.html
+++ b/app/index.html
@@ -1,25 +1,24 @@
-<!DOCTYPE html>
-<html>
-  <head>
+<!DOCTYPE html><html><head>
     <meta charset=\\"utf-8\\">
     <meta http-equiv=\\"X-UA-Compatible\\" content=\\"IE=edge\\">
-    <title>TestApp</title>
+    <title></title>
     <meta name=\\"description\\" content=\\"\\">
     <meta name=\\"viewport\\" content=\\"width=device-width, initial-scale=1\\">
 
-    {{content-for \\"head\\"}}
+    </head><body>{{content-for \\"head\\"}}
 
-    <link integrity=\\"\\" rel=\\"stylesheet\\" href=\\"{{rootURL}}assets/vendor.css\\">
+    <link integrity=\\"\\" rel=\\"stylesheet\\" href=\\"{{rootURL}}assets/tailwind.css\\"><link integrity=\\"\\" rel=\\"stylesheet\\" href=\\"{{rootURL}}assets/vendor.css\\">
     <link integrity=\\"\\" rel=\\"stylesheet\\" href=\\"{{rootURL}}assets/test-app.css\\">
 
     {{content-for \\"head-footer\\"}}
-  </head>
-  <body>
+  
+  
     {{content-for \\"body\\"}}
 
     <script src=\\"{{rootURL}}assets/vendor.js\\"></script>
     <script src=\\"{{rootURL}}assets/test-app.js\\"></script>
 
     {{content-for \\"body-footer\\"}}
-  </body>
-</html>
+  
+
+</body></html>
\\\\ No newline at end of file
diff --git a/package.json b/package.json
index 976763c..e14a6d4 100644
--- a/package.json
+++ b/package.json
@@ -11,7 +11,7 @@
     \\"test\\": \\"tests\\"
   },
   \\"scripts\\": {
-    \\"build\\": \\"ember build --environment=production\\",
+    \\"build\\": \\"npm run tailwind:build && ember build --environment=production\\",
     \\"lint\\": \\"npm-run-all --aggregate-output --continue-on-error --parallel \\\\\\"lint:!(fix)\\\\\\"\\",
     \\"lint:fix\\": \\"npm-run-all --aggregate-output --continue-on-error --parallel lint:*:fix\\",
     \\"lint:hbs\\": \\"ember-template-lint .\\",
@@ -20,7 +20,9 @@
     \\"lint:js:fix\\": \\"eslint . --fix\\",
     \\"start\\": \\"ember serve\\",
     \\"test\\": \\"npm-run-all lint test:*\\",
-    \\"test:ember\\": \\"ember test\\"
+    \\"test:ember\\": \\"ember test\\",
+    \\"tailwind:build\\": \\"npx tailwindcss -i ./tailwind-input.css -o ./public/assets/tailwind.css\\",
+    \\"tailwind:watch\\": \\"npx tailwindcss -i ./tailwind-input.css -o ./public/assets/tailwind.css --watch\\"
   },
   \\"devDependencies\\": {
     \\"@ember/optional-features\\": \\"^2.0.0\\",
@@ -59,7 +61,10 @@
     \\"prettier\\": \\"^2.5.1\\",
     \\"qunit\\": \\"^2.17.2\\",
     \\"qunit-dom\\": \\"^2.0.0\\",
-    \\"webpack\\": \\"^5.65.0\\"
+    \\"webpack\\": \\"^5.65.0\\",
+    \\"autoprefixer\\": \\"^10.0.0\\",
+    \\"postcss\\": \\"^8.0.0\\",
+    \\"tailwindcss\\": \\"^3.0.0\\"
   },
   \\"engines\\": {
     \\"node\\": \\"12.* || 14.* || >= 16\\"
diff --git a/tests/index.html b/tests/index.html
index ec6650a..b61f82e 100644
--- a/tests/index.html
+++ b/tests/index.html
@@ -1,23 +1,21 @@
-<!DOCTYPE html>
-<html>
-  <head>
+<!DOCTYPE html><html><head>
     <meta charset=\\"utf-8\\">
     <meta http-equiv=\\"X-UA-Compatible\\" content=\\"IE=edge\\">
-    <title>TestApp Tests</title>
+    <title></title>
     <meta name=\\"description\\" content=\\"\\">
     <meta name=\\"viewport\\" content=\\"width=device-width, initial-scale=1\\">
 
-    {{content-for \\"head\\"}}
+    </head><body>{{content-for \\"head\\"}}
     {{content-for \\"test-head\\"}}
 
-    <link rel=\\"stylesheet\\" href=\\"{{rootURL}}assets/vendor.css\\">
+    <link rel=\\"stylesheet\\" href=\\"{{rootURL}}assets/tailwind.css\\"><link rel=\\"stylesheet\\" href=\\"{{rootURL}}assets/vendor.css\\">
     <link rel=\\"stylesheet\\" href=\\"{{rootURL}}assets/test-app.css\\">
     <link rel=\\"stylesheet\\" href=\\"{{rootURL}}assets/test-support.css\\">
 
     {{content-for \\"head-footer\\"}}
     {{content-for \\"test-head-footer\\"}}
-  </head>
-  <body>
+  
+  
     {{content-for \\"body\\"}}
     {{content-for \\"test-body\\"}}
 
@@ -28,7 +26,7 @@
       </div>
     </div>
 
-    <script src=\\"/testem.js\\" integrity=\\"\\" data-embroider-ignore></script>
+    <script src=\\"/testem.js\\" integrity=\\"\\" data-embroider-ignore=\\"\\"></script>
     <script src=\\"{{rootURL}}assets/vendor.js\\"></script>
     <script src=\\"{{rootURL}}assets/test-support.js\\"></script>
     <script src=\\"{{rootURL}}assets/test-app.js\\"></script>
@@ -36,5 +34,6 @@
 
     {{content-for \\"body-footer\\"}}
     {{content-for \\"test-body-footer\\"}}
-  </body>
-</html>
+  
+
+</body></html>
\\\\ No newline at end of file"
`;

exports[`CLI > default command > using Tailwind as an example > local relative path 1`] = `
" .gitignore       |  3 +++
 app/index.html   | 19 +++++++++----------
 package.json     | 11 ++++++++---
 tests/index.html | 21 ++++++++++-----------
 4 files changed, 30 insertions(+), 24 deletions(-)"
`;

exports[`CLI > default command > using Tailwind as an example > local relative path 2`] = `
"diff --git a/.gitignore b/.gitignore
index 9bab45a..c4e2f3e 100644
--- a/.gitignore
+++ b/.gitignore
@@ -1,6 +1,9 @@
 # See https://help.github.com/ignore-files/ for more about ignoring files.
 
+
 # compiled output
+public/assets/tailwind.css
+
 /dist/
 /tmp/
 
diff --git a/app/index.html b/app/index.html
index 2398775..f038aae 100644
--- a/app/index.html
+++ b/app/index.html
@@ -1,25 +1,24 @@
-<!DOCTYPE html>
-<html>
-  <head>
+<!DOCTYPE html><html><head>
     <meta charset=\\"utf-8\\">
     <meta http-equiv=\\"X-UA-Compatible\\" content=\\"IE=edge\\">
-    <title>TestApp</title>
+    <title></title>
     <meta name=\\"description\\" content=\\"\\">
     <meta name=\\"viewport\\" content=\\"width=device-width, initial-scale=1\\">
 
-    {{content-for \\"head\\"}}
+    </head><body>{{content-for \\"head\\"}}
 
-    <link integrity=\\"\\" rel=\\"stylesheet\\" href=\\"{{rootURL}}assets/vendor.css\\">
+    <link integrity=\\"\\" rel=\\"stylesheet\\" href=\\"{{rootURL}}assets/tailwind.css\\"><link integrity=\\"\\" rel=\\"stylesheet\\" href=\\"{{rootURL}}assets/vendor.css\\">
     <link integrity=\\"\\" rel=\\"stylesheet\\" href=\\"{{rootURL}}assets/test-app.css\\">
 
     {{content-for \\"head-footer\\"}}
-  </head>
-  <body>
+  
+  
     {{content-for \\"body\\"}}
 
     <script src=\\"{{rootURL}}assets/vendor.js\\"></script>
     <script src=\\"{{rootURL}}assets/test-app.js\\"></script>
 
     {{content-for \\"body-footer\\"}}
-  </body>
-</html>
+  
+
+</body></html>
\\\\ No newline at end of file
diff --git a/package.json b/package.json
index 976763c..e14a6d4 100644
--- a/package.json
+++ b/package.json
@@ -11,7 +11,7 @@
     \\"test\\": \\"tests\\"
   },
   \\"scripts\\": {
-    \\"build\\": \\"ember build --environment=production\\",
+    \\"build\\": \\"npm run tailwind:build && ember build --environment=production\\",
     \\"lint\\": \\"npm-run-all --aggregate-output --continue-on-error --parallel \\\\\\"lint:!(fix)\\\\\\"\\",
     \\"lint:fix\\": \\"npm-run-all --aggregate-output --continue-on-error --parallel lint:*:fix\\",
     \\"lint:hbs\\": \\"ember-template-lint .\\",
@@ -20,7 +20,9 @@
     \\"lint:js:fix\\": \\"eslint . --fix\\",
     \\"start\\": \\"ember serve\\",
     \\"test\\": \\"npm-run-all lint test:*\\",
-    \\"test:ember\\": \\"ember test\\"
+    \\"test:ember\\": \\"ember test\\",
+    \\"tailwind:build\\": \\"npx tailwindcss -i ./tailwind-input.css -o ./public/assets/tailwind.css\\",
+    \\"tailwind:watch\\": \\"npx tailwindcss -i ./tailwind-input.css -o ./public/assets/tailwind.css --watch\\"
   },
   \\"devDependencies\\": {
     \\"@ember/optional-features\\": \\"^2.0.0\\",
@@ -59,7 +61,10 @@
     \\"prettier\\": \\"^2.5.1\\",
     \\"qunit\\": \\"^2.17.2\\",
     \\"qunit-dom\\": \\"^2.0.0\\",
-    \\"webpack\\": \\"^5.65.0\\"
+    \\"webpack\\": \\"^5.65.0\\",
+    \\"autoprefixer\\": \\"^10.0.0\\",
+    \\"postcss\\": \\"^8.0.0\\",
+    \\"tailwindcss\\": \\"^3.0.0\\"
   },
   \\"engines\\": {
     \\"node\\": \\"12.* || 14.* || >= 16\\"
diff --git a/tests/index.html b/tests/index.html
index ec6650a..b61f82e 100644
--- a/tests/index.html
+++ b/tests/index.html
@@ -1,23 +1,21 @@
-<!DOCTYPE html>
-<html>
-  <head>
+<!DOCTYPE html><html><head>
     <meta charset=\\"utf-8\\">
     <meta http-equiv=\\"X-UA-Compatible\\" content=\\"IE=edge\\">
-    <title>TestApp Tests</title>
+    <title></title>
     <meta name=\\"description\\" content=\\"\\">
     <meta name=\\"viewport\\" content=\\"width=device-width, initial-scale=1\\">
 
-    {{content-for \\"head\\"}}
+    </head><body>{{content-for \\"head\\"}}
     {{content-for \\"test-head\\"}}
 
-    <link rel=\\"stylesheet\\" href=\\"{{rootURL}}assets/vendor.css\\">
+    <link rel=\\"stylesheet\\" href=\\"{{rootURL}}assets/tailwind.css\\"><link rel=\\"stylesheet\\" href=\\"{{rootURL}}assets/vendor.css\\">
     <link rel=\\"stylesheet\\" href=\\"{{rootURL}}assets/test-app.css\\">
     <link rel=\\"stylesheet\\" href=\\"{{rootURL}}assets/test-support.css\\">
 
     {{content-for \\"head-footer\\"}}
     {{content-for \\"test-head-footer\\"}}
-  </head>
-  <body>
+  
+  
     {{content-for \\"body\\"}}
     {{content-for \\"test-body\\"}}
 
@@ -28,7 +26,7 @@
       </div>
     </div>
 
-    <script src=\\"/testem.js\\" integrity=\\"\\" data-embroider-ignore></script>
+    <script src=\\"/testem.js\\" integrity=\\"\\" data-embroider-ignore=\\"\\"></script>
     <script src=\\"{{rootURL}}assets/vendor.js\\"></script>
     <script src=\\"{{rootURL}}assets/test-support.js\\"></script>
     <script src=\\"{{rootURL}}assets/test-app.js\\"></script>
@@ -36,5 +34,6 @@
 
     {{content-for \\"body-footer\\"}}
     {{content-for \\"test-body-footer\\"}}
-  </body>
-</html>
+  
+
+</body></html>
\\\\ No newline at end of file"
`;
